Assets/MapGenerator.cs

```diff
@@ -0,0 +1,274 @@
+	private bool useRandomSeed = true;
+
+	[Range(0,100)]
+	public int randomCaveSeedPercent = 50;
+	[Range(0, 20)]
+	public int SmoothingIterations = 5;
+
+	public float heightScale = 1.0F;
+	public float xScale = 1.0F;
+	public float yNoise = 0f;
+	public float groundDepth = 50.0f;
+	public float hillDepth = 20.0f;
+	public float skyDepth = 20.0f;
+	public int height;
+
+	public AnimationCurve contourCurve = AnimationCurve.Linear (0f, 0f, 1f, 1f);
+	private float GroundTop;
+	public long heightSum;
+
+
+	[Range(0, 8)]
+	public int BirthLimit = 4;		//if the number of walls around the selected blank point is high enough then fill it in
+	[Range(0, 8)]
+	public int DeathLimit = 5;		//if the number of walls around the selected non-blank point is not too high then kill it
+
+
+	int[,] map;
+	int[,] newMap;
+	//MapCol[] map;
+	int[] Topography;
+
+	void Start() 
+	{
+		GenerateMap();
+	}
+
+	void Update() 
+	{
+		if (Input.GetMouseButtonDown(0)) 
+		{
+			GenerateMap();
+		}
+	}
+
+	void GenerateMap() 
+	{
+		height = (int)(groundDepth + hillDepth + skyDepth);
+		//map = new MapCol[width];
+		map = new int[width,height];
+		Topography = new int[width];
+		heightSum = 0;
+		ClearMap ();
+		//RandomFillMap();
+		GenerateHills ();
+		SeedCaves ();
+
+		for (int i = 0; i < SmoothingIterations; i ++) 
+		{
+			SmoothMap();
+		}
+	}
+
+	void GenerateHills()
+	{
+
+		if (useRandomSeed) 
+		{
+			seed = Time.time.ToString();
+		}
+		for(int x = 0;x < width;x++)
+		{
+			float xf = (float)x / (float)(width-1);
+			//height will evaluate to a value ~ between 0 and 1
+			GroundTop = contourCurve.Evaluate(xf) * heightScale * Mathf.PerlinNoise(xf * xScale, yNoise);
+			//multiply by hillDepth to get a realistic range of hills
+			GroundTop *= hillDepth;
+			//Add a solid chunk of ground underneath the hills
+			GroundTop += groundDepth;
+			//map[x].Topography = (int)GroundTop;
+			Topography[x] = (int)GroundTop;
+			//heightSum += map[x].Topography;
+			heightSum += Topography[x];
+			for(int y = 0;y < height;y++)
+			{
+				if(y <= GroundTop)
+				{
+					//map[x].Column[y] = 1;
+					map[x,y] = 1;
+				}
+				else
+				{
+					y = (int)height; 
+				}
+			}
+		}
+	}
+
+	void ClearMap()
+	{
+		for (int x = 0; x < width; x ++) 
+		{
+			for (int y = 0; y < height; y ++) 
+			{
+				map[x,y] = 0;
+				//map[x].Column[y] = 0;
+			}
+		}
+	}
+
+	void SwapMaps()
+	{
+		for (int x = 0; x < width; x ++) 
+		{
+			for (int y = 0; y < height; y ++) 
+			{
+				map[x,y] = newMap[x,y];
+			}
+		}
+	}
+
+	void SeedCaves()	//generate a number of 'holes' in the ground where caves could form later
+	{
+		if (useRandomSeed) 
+		{
+			seed = Time.time.ToString();
+		}
+		System.Random pseudoRandom = new System.Random(seed.GetHashCode());
+		for (int x = 1; x < width-1; x ++)  //Don't seed the edges
+		{
+			//for (int y = 0; y < map[x].Topography-1; y ++) //stops 1 block short of ground surface
+			for (int y = 3; y < Topography[x]-5; y ++) //stops 5 block short of ground surface
```

Comment:
Anywhere we're using a literal to define certain behavior we should constantize it at the top. It makes the code more readable and also allows us to change the value in one place instead of potentially all over the place.